src/miniscript/satisfy.rs

```diff
@@ -212,339 +238,521 @@ impl_tuple_satisfier!(&'a A, &'b B, &'c C, &'d D, &'e E, &'f F,);
-                    (None, None) => None,
-                    (Some(mut lsat), None) => {
-                        lsat.push(vec![1]);
-                        Some(lsat)
-                    }
-                    (None, Some(mut rsat)) => {
-                        rsat.push(vec![]);
-                        Some(rsat)
-                    }
-                    (Some(mut lsat), Some(mut rsat)) => {
-                        if l.ty.mall.safe && !r.ty.mall.safe {
-                            rsat.push(vec![]);
-                            Some(rsat)
-                        } else if !l.ty.mall.safe && r.ty.mall.safe {
-                            lsat.push(vec![1]);
-                            Some(lsat)
-                        } else {
-                            if satisfy_cost(&lsat) + 2 <= satisfy_cost(&rsat) + 1 {
-                                lsat.push(vec![1]);
-                                Some(lsat)
-                            } else {
-                                rsat.push(vec![]);
-                                Some(rsat)
-                            }
-                        }
-                    }
-                }
-            }
+                let l_sat = Self::satisfy(&l.node, stfr);
+                let r_sat = Self::satisfy(&r.node, stfr);
+                Self::minimum(
+                    Satisfaction {
+                        stack: Witness::combine(l_sat.stack, Witness::push_1()),
+                        has_sig: l_sat.has_sig,
+                    },
+                    Satisfaction {
+                        stack: Witness::combine(r_sat.stack, Witness::push_0()),
+                        has_sig: r_sat.has_sig,
+                    },
+                )
+            },
             Terminal::Thresh(k, ref subs) => {
-                fn flatten(v: Vec<Vec<Vec<u8>>>) -> Vec<Vec<u8>> {
-                    v.into_iter().fold(vec![], |mut acc, x| {
-                        acc.extend(x);
-                        acc
-                    })
-                }
+                let mut sats = subs
+                    .iter()
+                    .map(|s| Self::satisfy(&s.node, stfr))
+                    .collect::<Vec<_>>();
+                // Start with the to-return stack set to all dissatisfactions
+                let mut ret_stack = subs
+                    .iter()
+                    .map(|s| Self::satisfy(&s.node, stfr))
+                    .collect::<Vec<_>>();
+
+                // Sort everything by (sat cost - dissat cost), except that
+                // satisfactions without signatures beat satisfactions with
+                // signatures
+                let mut sat_indices = (0..subs.len()).collect::<Vec<_>>();
+                sat_indices.sort_by_key(|&i| {
+                    let stack_weight = match (&sats[i].stack, &ret_stack[i].stack) {
+                        (&Witness::Unavailable, _) => i64::MAX,
+                        (_, &Witness::Unavailable) => i64::MIN,
```

Comment:
Can this be unreachable.

Reply:
No, because `pk_h` is `d` but if the `Satisfier` doesn't know the preimage to the pubkey hash, dissatisfaction will fail nonetheless.